src/Features/LanguageServer/CommonLanguageServerProtocol.Framework/LanguageServerTarget.cs

```diff
@@ -0,0 +1,427 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.Contracts;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using StreamJsonRpc;
+
+namespace CommonLanguageServerProtocol.Framework;
+
+public interface ILspServiceProvider
+{
+    ILspServices CreateServices(string serverKind);
+}
+
+public interface ILspServices : IDisposable
+{
+    T GetRequiredService<T>();
+}
+
+public interface IRequestExecutionQueue<RequestContextType> where RequestContextType : struct
+{
+    Task<RequestContextType?> CreateRequestContextAsync(IQueueItem<RequestContextType> queueItem, CancellationToken cancellationToken);
+
+    event EventHandler<RequestShutdownEventArgs>? RequestServerShutdown;
+
+    Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(
+        bool mutatesSolutionState,
+        bool requiresLSPSolution,
+        IRequestHandler<TRequestType, TResponseType, RequestContextType> handler,
+        TRequestType request,
+        ClientCapabilities clientCapabilities,
+        string methodName,
+        CancellationToken cancellationToken);
+
+    void Shutdown();
+}
+
+public interface IQueueItem<RequestContextType> where RequestContextType : struct
+{
+    /// <summary>
+    /// Begins executing the work specified by this queue item.
+    /// </summary>
+    Task CallbackAsync(RequestContextType? context, CancellationToken cancellationToken);
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.RequiresLSPSolution" />
+    bool RequiresLSPSolution { get; }
+
+    /// <inheritdoc cref="IRequestHandler{RequestContextType}.MutatesSolutionState" />
+    bool MutatesSolutionState { get; }
+
+    string MethodName { get; }
+
+    /// <summary>
+    /// The document identifier that will be used to find the solution and document for this request. This comes from the <see cref="TextDocumentIdentifier"/> returned from the handler itself via a call to <see cref="IRequestHandler{RequestType, ResponseType, TResponseContextType}.GetTextDocumentIdentifier(RequestType)"/>.
+    /// </summary>
+    TextDocumentIdentifier? TextDocument { get; }
+
+    /// <inheritdoc cref="RequestContext.ClientCapabilities" />
+    ClientCapabilities ClientCapabilities { get; }
+
+    /// <summary>
+    /// <see cref="CorrelationManager.ActivityId"/> used to properly correlate this work with the loghub
+    /// tracing/logging subsystem.
+    /// </summary>
+    Guid ActivityId { get; }
+
+    IRequestMetrics Metrics { get; }
+}
+
+public interface IRequestMetrics
+{
+    void RecordCancellation();
```

Comment:
Notes for when we discuss this later - unsure on if this is the right abstraction - alternatives would be some kind event / notification when these things happen.  Maybe these need params on what happened here